mkosi/__init__.py

```diff
@@ -2993,9 +3003,12 @@ def run_shell(args: Args, config: Config) -> None:
         cmdline += [f"--set-credential={k}:{v}"]
 
     with contextlib.ExitStack() as stack:
+        # Make sure the latest nspawn settings are always used.
         if config.nspawn_settings:
+            exists = (config.output_dir_or_cwd() / f"{name}.nspawn").exists()
             shutil.copy2(config.nspawn_settings, config.output_dir_or_cwd() / f"{name}.nspawn")
-            stack.callback(lambda: (config.output_dir_or_cwd() / f"{name}.nspawn").unlink())
+            if not exists:
+                stack.callback(lambda: (config.output_dir_or_cwd() / f"{name}.nspawn").unlink())
```

Comment:
Since the callback is executed only later, this could be turned around
```
if not (config.output_dir_or_cwd() / f"{name}.nspawn").exists():
    stack.callback(lambda: (config.output_dir_or_cwd() / f"{name}.nspawn").unlink())
shutil.copy2(config.nspawn_settings, config.output_dir_or_cwd() / f"{name}.nspawn")
```
which I think is a bit clearer that this is setting up cleanup before copying it into place.

A missing_ok=True might also make this a bit more robust, since we probably don't care if it's already gone.